File: src/main/java/com/duxtinto/intellij/plugin/github/codereviews/domain/pullrequests/PullRequestEntity.java
package com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests;

import com.duxtinto.intellij.plugin.github.codereviews.domain.issues.IssueEntity;
import com.google.auto.value.AutoValue;
import com.google.common.base.Enums;

import java.util.List;

@AutoValue
public abstract class PullRequestEntity {
    public abstract long number();
    public abstract List<IssueEntity> closableIssues();
    public abstract String title();
    public abstract State state();
    public abstract String url();

    public static PullRequestEntity create(long number, List<IssueEntity> closableIssues, String title, State state, String url) {
        return new AutoValue_PullRequestEntity(number, closableIssues, title, state, url);
    }

    public enum State {
        UNKNOWN,
        OPEN,
        CLOSED;

        public static State fromString(String string) {
            return Enums.getIfPresent(State.class, string.toUpperCase()).or(State.UNKNOWN);
        }
    }
}

File: src/main/kotlin/com/duxtinto/intellij/plugin/github/codereviews/domain/pullrequests/PullRequestEntity.kt
package com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests

import com.duxtinto.intellij.plugin.github.codereviews.domain.issues.IssueEntity
import com.google.common.base.Enums
import java.net.URL

data class PullRequestEntity (
        val number: Long,
        val closeableIssues: List<IssueEntity> = listOf(),
        val title: String = "",
        val state: State = State.OPEN,
        val url: URL? = null,
        val head: Link? = null) {
    enum class State {
        UNKNOWN,
        OPEN,
        CLOSED;
        companion object {
            fun fromString(string: String): State {
                return Enums.getIfPresent(State::class.java, string.toUpperCase()).or(State.UNKNOWN)
            }
        }
    }

    data class Link(
            val label: String,
            val ref: String,
            val sha: String
    )
}
